zerver/views/auth.py

```diff
@@ -933,6 +919,96 @@ def api_fetch_api_key(
+
+
+@csrf_exempt
+@has_request_variables
+def jwt_fetch_api_key(
+    request: HttpRequest,
+    json_web_token: str = REQ(default=""),
+    include_profile: bool = REQ(default=False, json_validator=check_bool),
+) -> HttpResponse:
+    return_data: Dict[str, bool] = {}
+
+    realm = get_realm_from_request(request)
+    if realm is None:
+        raise InvalidSubdomainError()
+
+    try:
+        key = settings.JWT_AUTH_KEYS[realm.subdomain]["key"]
+        algorithms = settings.JWT_AUTH_KEYS[realm.subdomain]["algorithms"]
+    except KeyError:
+        raise JsonableError(_("Auth key for this subdomain not found"))
+
+    try:
+        json_web_token = request.POST["json_web_token"]
+        options = {"verify_signature": True}
+        payload = jwt.decode(json_web_token, key, algorithms=algorithms, options=options)
+    except KeyError:
+        raise JsonableError(_("No JSON web token passed in request"))
+    except jwt.InvalidTokenError:
+        raise JsonableError(_("Bad JSON web token"))
+
+    remote_email = payload.get("email", None)
+    if remote_email is None or not remote_email:
+        raise JsonableError(_("No email specified in JSON web token claims"))
+
+    user_profile = authenticate(
+        username=remote_email, realm=realm, return_data=return_data, use_dummy_backend=True
+    )
+    if user_profile is not None:
+        assert isinstance(user_profile, UserProfile)
+
+    api_key = process_api_key_fetch_authenticate_result(request, user_profile, return_data)
+    assert user_profile is not None
+
+    result: Dict[str, Any] = {
+        "api_key": api_key,
+        "email": user_profile.delivery_email,
+    }
+
+    if include_profile:
+        members = get_raw_user_data(
+            realm,
+            user_profile,
+            target_user=user_profile,
+            client_gravatar=False,
+            user_avatar_url_field_optional=False,
+            include_custom_profile_fields=False,
+        )
+        result["user"] = members[user_profile.id]
+
+    return json_success(request, data=result)
+
+
+@csrf_exempt
+@require_post
+@has_request_variables
+def api_fetch_api_key(
+    request: HttpRequest, username: str = REQ(), password: str = REQ()
+) -> HttpResponse:
+    return_data: Dict[str, bool] = {}
+
+    realm = get_realm_from_request(request)
+    if realm is None:
+        raise InvalidSubdomainError()
+
+    if not ldap_auth_enabled(realm=realm):
+        # In case we don't authenticate against LDAP, check for a valid
+        # email. LDAP backend can authenticate against a non-email.
+        validate_login_email(username)
+    user_profile = authenticate(
+        request=request, username=username, password=password, realm=realm, return_data=return_data
+    )
+    if user_profile is not None:
+        assert isinstance(user_profile, UserProfile)
+
+    api_key = api_key = process_api_key_fetch_authenticate_result(
+        request, user_profile, return_data
+    )
+    assert user_profile is not None
```

Comment:
I don't understand this bit -- I think we're likely missing an exit path for this function if `user_profile is None` -- i.e. a failed authentication.

Reply:
Oh, I see, `process_api_key_fetch_authenticate_result` raises an exception in that case. That deserves a comment at least; there's probably a fancy type assertion we could declare on `process_api_key_fetch_authenticate_result` as well.

But maybe a better plan would be to have a function that we call if `user_profile` is None called `process_api_key_fetch_authenticate_failure`, and have the logic to raise the appropriate exception be there; or maybe even `raise get_api_key_fetch_authenticaiton_exception(return_data)` (with a `return` rather than `raise` as what that function returns).

Reply:
Yeah, that looks good.